### PR TITLE
docs: Add documentation based on the youtube video

### PR DESCRIPTION
**Pull request by joshka**
There's a lot of technical detail that went into creating this CI configuration, and you noted that it wasn't documented particularly well in the youtube vid covering it. I've added my understanding of what you wrote, as well as some basic instructions on how to incorporate this into an existing repo. I took the approach of what would I need to understand if looking at each action / workflow just by reading the code (rather than having to jump off into PRs and twitter messages to understand). I added a readme to the .github folder, which I think would be useful to help guide users into a pit of success using this repo.

Thanks for creating this. It's pretty useful.

### CODE REVIEW

File: .github/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+reasonably sensible base for writing your own ci on. By using this strategy
+the history of the CI repo is included in your repo, and future updates to
+the CI can be merged later.
+
+To perform this merge run:
+
+```shell
+git remote add ci https://github.com/jonhoo/rust-ci-conf.git
+git fetch ci
+git merge --allow-unrelated-histories
```
Comment (kovyrin) [inline .github/README.md]:
```suggestion
git merge --allow-unrelated-histories ci/main
```

**Comment by kovyrin** [comment]
Noticed an issues with this PR. The README.md file in .github directory ends up overriding the main readme file used by Github web UI (the UI picks up this file instead of the one in the root of the project). I think it'd make sense to rename it into something else.

**Comment by joshka** [comment]
@kovyrin fixed the issue with the naming.

File: .github/workflows/check.yml [hunk 2]
```diff
@@ -71,9 +82,11 @@ jobs:
       - name: cargo install cargo-hack
         uses: taiki-e/install-action@cargo-hack
       # intentionally no target specifier; see https://github.com/jonhoo/rust-ci-conf/pull/4
+      # --feature-powerset runs for every combination of features
       - name: cargo hack
         run: cargo hack --feature-powerset check
   msrv:
+    # determine the minimal rust version that is supported by this crate
```
Comment (jonhoo) [inline .github/workflows/check.yml]: Ah, this isn't quite right: this doesn't _determine_ the MSRV, it checks whether the configured MSRV can indeed still build the current crate. So it'll warn you if you accidentally bump the MSRV, but it will not _find_ the MSRV in such a case.

File: .github/workflows/nostd.yml [hunk 3]
```diff
@@ -1,10 +1,14 @@
+# This workflow checks whether the library is able to run without the std library (e.g. embedded).
+# This entire file should be removed if this crate does not support no-std
```
Comment (jonhoo) [inline .github/workflows/nostd.yml]: Maybe also add a line here (and in the other non-`check` files) saying that all the shared configuration lines (like `on: push:`) are documented in `check.yml`, and then remove the repeated comments (like "If new code is pushed") from all the other files.

File: .github/workflows/scheduled.yml [hunk 4]
```diff
@@ -29,6 +33,10 @@ jobs:
         run: cargo test --locked --all-features --all-targets
   # https://twitter.com/alcuadrado/status/1571291687837732873
   update:
+    # This action checks that updating the dependencies of this crate to the
+    # latest available that satisfy the versions in Cargo.toml does not break
+    # this crate. This is important as consumers of this crate will generally
+    # use the latest available crates.
```
Comment (jonhoo) [inline .github/workflows/scheduled.yml]: > subject to the constraints in our `Cargo.toml` (i.e., they will not update to `2.0` if we specify `1.3`).

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -4,7 +4,9 @@ on:
   push:
     branches: [main]
   pull_request:
-# Spend CI time only on latest ref: https://github.com/jonhoo/rust-ci-conf/pull/5
+# If new code is pushed to a PR branch, then cancel in progress workflows for that PR.
```
Comment (jonhoo) [inline .github/workflows/test.yml]: This file is missing top-level documentation.

File: .github/workflows/check.yml [hunk 6]
```diff
@@ -1,10 +1,15 @@
 permissions:
   contents: read
+# This configuration allows maintainers of this repo to create a branch and pull request based on
+# the new branch. Restricting the push trigger to the main branch ensures that the PR only gets
+# built once.
```
Comment (jonhoo) [inline .github/workflows/check.yml:12]: This file is missing top-level documentation.

File: .github/workflows/test.yml [hunk 7]
```diff
@@ -34,6 +39,17 @@ jobs:
       - name: cargo test --doc
         run: cargo test --locked --all-features --doc
   minimal:
+    # This action chooses the oldest version of the dependencies to ensure that this crate is
+    # compatible with the minimal version that this crate and its dependencies require.
+    # This will pickup issues where this create relies on functionality that was introduced
+    # later than the actual version specified (e.g. when we choose just a major version, but a
```
Comment (jonhoo) [inline .github/workflows/test.yml]: nit: "e.g." and "i.e." should basically always be followed by a comma (according to _most_ style guides).

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -34,6 +39,17 @@ jobs:
       - name: cargo test --doc
         run: cargo test --locked --all-features --doc
   minimal:
+    # This action chooses the oldest version of the dependencies to ensure that this crate is
```
Comment (jonhoo) [inline .github/workflows/test.yml]: > oldest version of dependencies permitted by `Cargo.toml`

File: .github/workflows/test.yml [hunk 9]
```diff
@@ -34,6 +39,17 @@ jobs:
...
+    # compatible with the minimal version that this crate and its dependencies require.
+    # This will pickup issues where this create relies on functionality that was introduced
+    # later than the actual version specified (e.g. when we choose just a major version, but a
+    # method was added after this version).
+    #
+    # This particular check can be difficult to get to succeed as often transitive dependencies
+    # may be incorrectly specified (e.g. a dependency specifies 1.0 but really requires 1.1.5).
+    # There is an alternative flag available -Zminimal-direct that uses the minimal versions for
+    # direct dependencies of this crate, while selecting the maximal versions for the transitive
+    # dependencies
```
Comment (jonhoo) [inline .github/workflows/test.yml]: > Alternatively, you can add a line in your `Cargo.toml` to artificially increase the minimal dependency, which you do with:
> ```toml
> # for minimal-versions
> [target.'cfg(any())'.dependencies]
> openssl = { version = "0.10.55", optional = true } # needed to allow foo to build with -Zminimal-versions
> ```

The `optional = true` is necessary in case that dependency isn't otherwise transitively required by your library, and the `target` bit is so that this dependency edge never actually affects Cargo build order.

See also https://github.com/jonhoo/fantoccini/blob/fde336472b712bc7ebf5b4e772023a7ba71b2262/Cargo.toml#L47-L49.

**Review by jonhoo — CHANGES_REQUESTED** [review]
This is great, thanks for taking the time! Left a few notes inline.

**Comment by c-git** [comment]
# TLDR 

I want to make a PR that starts from this one and adds what I learn during testing and want to know if @joshka would be ok with that?

# Long Version

@joshka Do you have the time to make the requested changes? I'm trying to integrate this repo into my workflow and want to have the headings you were adding.  I was in the process of testing how these work and wanted a place to document what I find. Going to defer testing for now until this PR lands. I realistically don't know when I will be able to get back to this (just didn't want to duplicate work now nor step on anyone's toes) but if you haven't had time to do it yet before I get back to it would you mind me creating a "follow up" PR that starts from you PR and add what I learn?

**Comment by joshka** [comment]
@c-git apologies for not getting to this sooner. Just going through a bunch of PRs to clean up things now - I've addressed All @jonhoo's comments with the latest push.

**Comment by c-git** [comment]
No worries, I know how life goes.

File: .github/workflows/nostd.yml [hunk 10]
```diff
@@ -1,10 +1,12 @@
+# This workflow checks whether the library is able to run without the std library (e.g., embedded).
+# This entire file should be removed if this crate does not support no-std See check.yml for
```
Comment (jonhoo) [inline .github/workflows/nostd.yml]:
```suggestion
# This entire file should be removed if this crate does not support no-std. See check.yml for
```

File: .github/workflows/scheduled.yml [hunk 11]
```diff
@@ -29,12 +31,16 @@ jobs:
...
+    # to).
     runs-on: ubuntu-latest
     name: ubuntu / beta / updated
-    # There's no point running this if no Cargo.lock was checked in in the
-    # first place, since we'd just redo what happened in the regular test job.
-    # Unfortunately, hashFiles only works in if on steps, so we reepeat it.
-    # if: hashFiles('Cargo.lock') != ''
+    # There's no point running this if no Cargo.lock was checked in in the first place, since we'd
+    # just redo what happened in the regular test job. Unfortunately, hashFiles only works in if on
+    # steps, so we repeat it. if: hashFiles('Cargo.lock') != ''
```
Comment (jonhoo) [inline .github/workflows/scheduled.yml]:
```suggestion
    # steps, so we repeat it.
```

**Review by jonhoo — APPROVED** [review]
Looking great, thanks!
